### PR TITLE
[WIP] Add ISO/IEC Base Media File Format

### PR DESCRIPTION
**Pull request by 1div0**
These are the very first steps toward bringing support of the ISO/IEC Base Media File Format including AVIF, HEIF and CR3.

Please review.

Many thanks beforehand.

### CODE REVIEW

**Comment by clanmills** [comment]
This looks like a good start, Peter.  Your code is causing the existing JP2 code in src/jp2image.cpp to be ignored and instead throws the error "The file contains data of an unknown image type" which causes every reference to that code by the test suite to fail.  This is in the Image Factory and isn't very difficult to fix, so I'll fix this for you and get your PR to pass the test suite.

The ImageFactory sequentially searches a table of image types.  We have to be sure to call the jp2image.cpp "sniffer" before your nice new src/isobmff.cpp code.  There's a bug in that table concerning jp2 which I've changed to 19 (in jp2image.cpp).  Those const int things have been changed to an enum in 'master'.  We have to endure the pain that 

Can you change isobmff.cpp/.hpp to be isoimage.cpp/.hpp to match the other image handlers jp2image, jpegimage and so on. 

Once we have the "boiler plate" in place, we'll have to focus on readMetadata() and that involves porting the code from tvisitor class Jp2Image into exiv2.  And then update the test suite and documentation README.md and man/man1/exiv2.1.  We'll get there soon enough.

**Comment by clanmills** [comment]
Several comments, @1div0 

1. Maybe we should remove all mention of ISO and simply refer to this feature as BMFF.  It's in the branch name, however that's temporary scaffolding which will disappear.
 
2. I suspect `static bool isBigEndian()` is a relic and shouldn't be in src/jp2image.cpp (or bmffimage.cpp).  We can use Image::isBigEndianPlatform() instead.  If you don't want to touch src/jp2image.cpp, that's fine.

3. Can you add code to version.cpp to output enable_bmff:
```bash
733 rmills@rmillsmbp:~/gnu/github/exiv2/ISOBMFF/build $ bin/exiv2 -vV -g enable
exiv2 0.27.4.9
enable_video=0
enable_webready=0
enable_nls=0
734 rmills@rmillsmbp:~/gnu/github/exiv2/ISOBMFF/build $ 
```

Not only is this useful for inspection, the test harness is very likely to inspect it and skip tests when this bmff is not enabled.

**Comment by 1div0** [comment]
1. Sure, I will remove ISO abbreviation, the International Organization for Standardization. The feudal organization, as Leonardo mentioned.

2. Most likely. I experimented with use of preprocessor macros as lowest overhead option.

3. Will do so.

File: src/bmffimage.cpp [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+            std::cout << "Exiv2::BmffImage::readMetadata: "
+                      << "Position: " << position
+                      << " box type: " << toAscii(type)
+                      << " length: " << length
+                      << std::endl;
+#endif
+
+            if (length == 0) return ;
+
+            if (length == 1)
```
Comment (hassec) [inline src/bmffimage.cpp:215]: Is it possible that this is missing something inside the curly braces?

Comment (clanmills) [reply src/bmffimage.cpp:215]: @hassec This code is very much a "boiler" plate image handler.  Although it recognises some file formats, it doesn't parse the file in readMetadata() to locate the Exiv2/IPTC/XMP/ICC data blocks.  The code to do this is in tvisitor.cpp in my book and has to be ported here.  Another matter is to implement writeMetadata() which serialises the _in-memory_ metadata into blocks and updates the file.  I think the necessary code will be of the order of 500 lines.

exiv2arch

@1div0 I'm concerned that you haven't updated the PR for a few weeks.  The project schedule for Code Complete is less than 3 weeks away on 2021-02-28.  Are you confident of making the schedule or would you like me to get involved?  https://github.com/Exiv2/exiv2/issues/1018#issuecomment-757979745

**Comment by clanmills** [comment]
I'm going to close this.  Work is proceeding here: #1475

**Comment by clanmills** [comment]
@1div0 Can delete this.  I think it's dead.

**Comment by 1div0** [comment]
Deleting now.

**Pull request closed without merging**